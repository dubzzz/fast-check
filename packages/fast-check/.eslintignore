bundle/
node_modules/
lib/
lib-*/
perf/
jest*.js
jest*.cjs