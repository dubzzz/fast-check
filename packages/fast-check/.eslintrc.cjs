module.exports = {
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'prettier',
  ],
  rules: {
    '@typescript-eslint/explicit-module-boundary-types': 'error',
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/interface-name-prefix': 'off',
    '@typescript-eslint/no-empty-function': 'warn',
    '@typescript-eslint/no-empty-interface': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/no-this-alias': 'warn',
    '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
    '@typescript-eslint/no-use-before-define': 'warn',
  },
  overrides: [
    {
      files: ['*.spec.ts', 'test/unit/**/*.ts', 'test/e2e/**/*.ts'],
      rules: {
        '@typescript-eslint/no-empty-function': 'off',
        '@typescript-eslint/no-use-before-define': 'off',
        '@typescript-eslint/no-non-null-assertion': 'off',
      },
    },
  ],
};
