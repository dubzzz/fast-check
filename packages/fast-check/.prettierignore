bundle/
coverage/
docs/
documentation/Arbitraries.md
lib/
lib-*/