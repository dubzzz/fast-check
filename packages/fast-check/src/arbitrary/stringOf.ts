import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import type { ArrayConstraintsInternal } from './array';
import { array } from './array';
import type { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { patternsToStringMapper, patternsToStringUnmapperFor } from './_internals/mappers/PatternsToString';
import { createSlicesForString } from './_internals/helpers/SlicesForStringBuilder';
export type { StringSharedConstraints } from './_shared/StringSharedConstraints';

const safeObjectAssign = Object.assign;

/**
 * For strings using the characters produced by `charArb`
 *
 * @param charArb - Arbitrary able to generate random strings (possibly multiple characters)
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 1.1.3
 * @public
 */
export function stringOf(charArb: Arbitrary<string>, constraints: StringSharedConstraints = {}): Arbitrary<string> {
  const unmapper = patternsToStringUnmapperFor(charArb, constraints);
  const experimentalCustomSlices = createSlicesForString(charArb, unmapper);
  // TODO - Move back to object spreading as soon as we bump support from es2017 to es2018+
  const enrichedConstraints: ArrayConstraintsInternal<string> = safeObjectAssign(safeObjectAssign({}, constraints), {
    experimentalCustomSlices,
  });
  return array(charArb, enrichedConstraints).map(patternsToStringMapper, unmapper);
}
