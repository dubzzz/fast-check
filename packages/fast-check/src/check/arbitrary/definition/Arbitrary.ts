import type { Random } from '../../../random/generator/Random';
import { Stream } from '../../../stream/Stream';
import { cloneMethod, hasCloneMethod } from '../../symbols';
import { Value } from './Value';

const safeObjectAssign = Object.assign;

/**
 * Abstract class able to generate values on type `T`
 *
 * The values generated by an instance of Arbitrary can be previewed - with {@link sample} - or classified - with {@link statistics}.
 *
 * @remarks Since 0.0.7
 * @public
 */
export abstract class Arbitrary<T> {
  /**
   * Generate a value of type `T` along with its context (if any)
   * based on the provided random number generator
   *
   * @param mrng - Random number generator
   * @param biasFactor - If taken into account 1 value over biasFactor must be biased. Either integer value greater or equal to 2 (bias) or undefined (no bias)
   * @returns Random value of type `T` and its context
   *
   * @remarks Since 0.0.1 (return type changed in 3.0.0)
   */
  abstract generate(mrng: Random, biasFactor: number | undefined): Value<T>;

  /**
   * Check if a given value could be pass to `shrink` without providing any context.
   *
   * In general, `canShrinkWithoutContext` is not designed to be called for each `shrink` but rather on very special cases.
   * Its usage must be restricted to `canShrinkWithoutContext` or in the rare* contexts of a `shrink` method being called without
   * any context. In this ill-formed case of `shrink`, `canShrinkWithoutContext` could be used or called if needed.
   *
   * *we fall in that case when fast-check is asked to shrink a value that has been provided manually by the user,
   *  in other words: a value not coming from a call to `generate` or a normal `shrink` with context.
   *
   * @param value - Value to be assessed
   * @returns `true` if and only if the value could have been generated by this instance
   *
   * @remarks Since 3.0.0
   */
  abstract canShrinkWithoutContext(value: unknown): value is T;

  /**
   * Shrink a value of type `T`, may rely on the context previously provided to shrink efficiently
   *
   * Must never be called with possibly invalid values and no context without ensuring that such call is legal
   * by calling `canShrinkWithoutContext` first on the value.
   *
   * @param value - The value to shrink
   * @param context - Its associated context (the one returned by generate) or `undefined` if no context but `canShrinkWithoutContext(value) === true`
   * @returns Stream of shrinks for value based on context (if provided)
   *
   * @remarks Since 3.0.0
   */
  abstract shrink(value: T, context: unknown | undefined): Stream<Value<T>>;

  /**
   * Create another arbitrary by filtering values against `predicate`
   *
   * All the values produced by the resulting arbitrary
   * satisfy `predicate(value) == true`
   *
   * Be aware that using filter may highly impact the time required to generate a valid entry
   *
   * @example
   * ```typescript
   * const integerGenerator: Arbitrary<number> = ...;
   * const evenIntegerGenerator: Arbitrary<number> = integerGenerator.filter(e => e % 2 === 0);
   * // new Arbitrary only keeps even values
   * ```
   *
   * @param refinement - Predicate, to test each produced element. Return true to keep the element, false otherwise
   * @returns New arbitrary filtered using predicate
   *
   * @remarks Since 1.23.0
   */
  filter<U extends T>(refinement: (t: T) => t is U): Arbitrary<U>;
  /**
   * Create another arbitrary by filtering values against `predicate`
   *
   * All the values produced by the resulting arbitrary
   * satisfy `predicate(value) == true`
   *
   * Be aware that using filter may highly impact the time required to generate a valid entry
   *
   * @example
   * ```typescript
   * const integerGenerator: Arbitrary<number> = ...;
   * const evenIntegerGenerator: Arbitrary<number> = integerGenerator.filter(e => e % 2 === 0);
   * // new Arbitrary only keeps even values
   * ```
   *
   * @param predicate - Predicate, to test each produced element. Return true to keep the element, false otherwise
   * @returns New arbitrary filtered using predicate
   *
   * @remarks Since 0.0.1
   */
  filter(predicate: (t: T) => boolean): Arbitrary<T>;
  filter<U extends T>(refinement: (t: T) => t is U): Arbitrary<U> {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new FilterArbitrary(this, refinement);
  }

  /**
   * Create another arbitrary by mapping all produced values using the provided `mapper`
   * Values produced by the new arbitrary are the result of applying `mapper` value by value
   *
   * @example
   * ```typescript
   * const rgbChannels: Arbitrary<{r:number,g:number,b:number}> = ...;
   * const color: Arbitrary<string> = rgbChannels.map(ch => `#${(ch.r*65536 + ch.g*256 + ch.b).toString(16).padStart(6, '0')}`);
   * // transform an Arbitrary producing {r,g,b} integers into an Arbitrary of '#rrggbb'
   * ```
   *
   * @param mapper - Map function, to produce a new element based on an old one
   * @param unmapper - Optional unmap function, it will never be used except when shrinking user defined values. Must throw if value is not compatible (since 3.0.0)
   * @returns New arbitrary with mapped elements
   *
   * @remarks Since 0.0.1
   */
  map<U>(mapper: (t: T) => U, unmapper?: (possiblyU: unknown) => T): Arbitrary<U> {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new MapArbitrary(this, mapper, unmapper);
  }

  /**
   * Create another arbitrary by mapping a value from a base Arbirary using the provided `fmapper`
   * Values produced by the new arbitrary are the result of the arbitrary generated by applying `fmapper` to a value
   * @example
   * ```typescript
   * const arrayAndLimitArbitrary = fc.nat().chain((c: number) => fc.tuple( fc.array(fc.nat(c)), fc.constant(c)));
   * ```
   *
   * @param chainer - Chain function, to produce a new Arbitrary using a value from another Arbitrary
   * @returns New arbitrary of new type
   *
   * @remarks Since 1.2.0
   */
  chain<U>(chainer: (t: T) => Arbitrary<U>): Arbitrary<U> {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new ChainArbitrary(this, chainer);
  }

  /**
   * Create another Arbitrary with no shrink values
   *
   * @example
   * ```typescript
   * const dataGenerator: Arbitrary<string> = ...;
   * const unshrinkableDataGenerator: Arbitrary<string> = dataGenerator.noShrink();
   * // same values no shrink
   * ```
   *
   * @returns Create another arbitrary with no shrink values
   * @remarks Since 0.0.9
   */
  noShrink(): Arbitrary<T> {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new NoShrinkArbitrary(this);
  }

  /**
   * Create another Arbitrary that cannot be biased
   *
   * @param freq - The biased version will be used one time over freq - if it exists
   * @remarks Since 1.1.0
   */
  noBias(): Arbitrary<T> {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new NoBiasArbitrary(this);
  }
}

/** @internal */
type ChainArbitraryContext<T, U> = {
  originalBias: number | undefined;
  originalValue: T;
  originalContext: unknown;
  stoppedForOriginal: boolean;
  chainedArbitrary: Arbitrary<U>;
  chainedContext: unknown;
  clonedMrng: Random;
};

/** @internal */
class ChainArbitrary<T, U> extends Arbitrary<U> {
  constructor(
    readonly arb: Arbitrary<T>,
    readonly chainer: (t: T) => Arbitrary<U>,
  ) {
    super();
  }
  generate(mrng: Random, biasFactor: number | undefined): Value<U> {
    const clonedMrng = mrng.clone();
    const src = this.arb.generate(mrng, biasFactor);
    return this.valueChainer(src, mrng, clonedMrng, biasFactor);
  }
  canShrinkWithoutContext(value: unknown): value is U {
    // TODO Need unchainer
    return false;
  }
  shrink(value: U, context?: unknown): Stream<Value<U>> {
    if (this.isSafeContext(context)) {
      return (
        !context.stoppedForOriginal
          ? this.arb
              .shrink(context.originalValue, context.originalContext)
              .map((v) => this.valueChainer(v, context.clonedMrng.clone(), context.clonedMrng, context.originalBias))
          : Stream.nil<Value<U>>()
      ).join(
        context.chainedArbitrary.shrink(value, context.chainedContext).map((dst) => {
          // TODO - Move back to object spreading as soon as we bump support from es2017 to es2018+
          const newContext: ChainArbitraryContext<T, U> = safeObjectAssign(safeObjectAssign({}, context), {
            chainedContext: dst.context,
            stoppedForOriginal: true,
          });
          return new Value(dst.value_, newContext);
        }),
      );
    }
    // TODO Need unchainer
    return Stream.nil();
  }
  private valueChainer(
    v: Value<T>,
    generateMrng: Random,
    clonedMrng: Random,
    biasFactor: number | undefined,
  ): Value<U> {
    const chainedArbitrary = this.chainer(v.value_);
    const dst = chainedArbitrary.generate(generateMrng, biasFactor);
    const context: ChainArbitraryContext<T, U> = {
      originalBias: biasFactor,
      originalValue: v.value_,
      originalContext: v.context,
      stoppedForOriginal: false,
      chainedArbitrary,
      chainedContext: dst.context,
      clonedMrng,
    };
    return new Value(dst.value_, context);
  }
  private isSafeContext(context: unknown): context is ChainArbitraryContext<T, U> {
    return (
      context != null &&
      typeof context === 'object' &&
      'originalBias' in (context as any) &&
      'originalValue' in (context as any) &&
      'originalContext' in (context as any) &&
      'stoppedForOriginal' in (context as any) &&
      'chainedArbitrary' in (context as any) &&
      'chainedContext' in (context as any) &&
      'clonedMrng' in (context as any)
    );
  }
}

/** @internal */
type MapArbitraryContext<T> = {
  originalValue: T;
  originalContext: unknown;
};

/** @internal */
class MapArbitrary<T, U> extends Arbitrary<U> {
  readonly bindValueMapper: (v: Value<T>) => Value<U>;
  constructor(
    readonly arb: Arbitrary<T>,
    readonly mapper: (t: T) => U,
    readonly unmapper?: (possiblyU: unknown) => T,
  ) {
    super();
    this.bindValueMapper = (v: Value<T>): Value<U> => this.valueMapper(v);
  }
  generate(mrng: Random, biasFactor: number | undefined): Value<U> {
    const g = this.arb.generate(mrng, biasFactor);
    return this.valueMapper(g);
  }
  canShrinkWithoutContext(value: unknown): value is U {
    if (this.unmapper !== undefined) {
      try {
        const unmapped = this.unmapper(value);
        return this.arb.canShrinkWithoutContext(unmapped);
      } catch (_err) {
        return false;
      }
    }
    return false;
  }
  shrink(value: U, context?: unknown): Stream<Value<U>> {
    if (this.isSafeContext(context)) {
      return this.arb.shrink(context.originalValue, context.originalContext).map(this.bindValueMapper);
    }
    if (this.unmapper !== undefined) {
      const unmapped = this.unmapper(value);
      // As `shrink` should never be called without a valid context
      // except if `canShrinkWithoutContext` tells that the value was compatible with a shrink without any context
      // we can safely consider `this.arb.canShrinkWithoutContext(unmapped)` to be true at that point.
      return this.arb.shrink(unmapped, undefined).map(this.bindValueMapper);
    }
    return Stream.nil();
  }
  private mapperWithCloneIfNeeded(v: Value<T>): [U, T] {
    const sourceValue = v.value;
    const mappedValue = this.mapper(sourceValue);
    if (
      v.hasToBeCloned &&
      ((typeof mappedValue === 'object' && mappedValue !== null) || typeof mappedValue === 'function') &&
      Object.isExtensible(mappedValue) &&
      !hasCloneMethod(mappedValue)
    ) {
      // WARNING: In case the mapped value is not extensible it will not be extended
      Object.defineProperty(mappedValue, cloneMethod, { get: () => () => this.mapperWithCloneIfNeeded(v)[0] });
    }
    return [mappedValue, sourceValue];
  }
  private valueMapper(v: Value<T>): Value<U> {
    const [mappedValue, sourceValue] = this.mapperWithCloneIfNeeded(v);
    const context: MapArbitraryContext<T> = { originalValue: sourceValue, originalContext: v.context };
    return new Value(mappedValue, context);
  }
  private isSafeContext(context: unknown): context is MapArbitraryContext<T> {
    return (
      context != null &&
      typeof context === 'object' &&
      'originalValue' in (context as any) &&
      'originalContext' in (context as any)
    );
  }
}

/** @internal */
class FilterArbitrary<T, U extends T> extends Arbitrary<U> {
  readonly bindRefinementOnValue: (v: Value<T>) => v is Value<U>;
  constructor(
    readonly arb: Arbitrary<T>,
    readonly refinement: (t: T) => t is U,
  ) {
    super();
    this.bindRefinementOnValue = (v: Value<T>): v is Value<U> => this.refinementOnValue(v);
  }
  generate(mrng: Random, biasFactor: number | undefined): Value<U> {
    // eslint-disable-next-line no-constant-condition
    while (true) {
      const g = this.arb.generate(mrng, biasFactor);
      if (this.refinementOnValue(g)) {
        return g;
      }
    }
  }
  canShrinkWithoutContext(value: unknown): value is U {
    return this.arb.canShrinkWithoutContext(value) && this.refinement(value);
  }
  shrink(value: U, context?: unknown): Stream<Value<U>> {
    return this.arb.shrink(value, context).filter(this.bindRefinementOnValue);
  }
  private refinementOnValue(v: Value<T>): v is Value<U> {
    return this.refinement(v.value);
  }
}

/** @internal */
class NoShrinkArbitrary<T> extends Arbitrary<T> {
  constructor(readonly arb: Arbitrary<T>) {
    super();
  }
  generate(mrng: Random, biasFactor: number | undefined): Value<T> {
    return this.arb.generate(mrng, biasFactor);
  }
  canShrinkWithoutContext(value: unknown): value is T {
    return this.arb.canShrinkWithoutContext(value);
  }
  shrink(_value: T, _context?: unknown): Stream<Value<T>> {
    return Stream.nil();
  }
  noShrink() {
    return this;
  }
}

/** @internal */
class NoBiasArbitrary<T> extends Arbitrary<T> {
  constructor(readonly arb: Arbitrary<T>) {
    super();
  }
  generate(mrng: Random, _biasFactor: number | undefined): Value<T> {
    return this.arb.generate(mrng, undefined);
  }
  canShrinkWithoutContext(value: unknown): value is T {
    return this.arb.canShrinkWithoutContext(value);
  }
  shrink(value: T, context?: unknown): Stream<Value<T>> {
    return this.arb.shrink(value, context);
  }
  noBias() {
    return this;
  }
}

/**
 * Ensure an instance is an instance of Arbitrary
 * @param instance - The instance to be checked
 * @internal
 */
export function isArbitrary(instance: unknown): instance is Arbitrary<unknown> {
  return (
    typeof instance === 'object' &&
    instance !== null &&
    'generate' in instance &&
    'shrink' in instance &&
    'canShrinkWithoutContext' in instance
  );
}

/**
 * Ensure an instance is an instance of Arbitrary
 * @param instance - The instance to be checked
 * @internal
 */
export function assertIsArbitrary(instance: unknown): asserts instance is Arbitrary<unknown> {
  if (!isArbitrary(instance)) {
    throw new Error('Unexpected value received: not an instance of Arbitrary');
  }
}
