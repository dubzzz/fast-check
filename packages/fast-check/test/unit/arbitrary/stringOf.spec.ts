import * as fc from 'fast-check';
import { stringOf } from '../../../src/arbitrary/stringOf';

import { Arbitrary } from '../../../src/check/arbitrary/definition/Arbitrary';
import { Value } from '../../../src/check/arbitrary/definition/Value';
import type { Random } from '../../../src/random/generator/Random';
import { Stream } from '../../../src/stream/Stream';
import {
  assertProduceSameValueGivenSameSeed,
  assertProduceValuesShrinkableWithoutContext,
} from './__test-helpers__/ArbitraryAssertions';
import { buildShrinkTree, renderTree } from './__test-helpers__/ShrinkTree';

describe('stringOf (integration)', () => {
  type Extra = { minLength?: number; maxLength?: number; patterns: string[] };
  const extraParameters: fc.Arbitrary<Extra> = fc
    .tuple(
      fc.nat({ max: 5 }),
      fc.nat({ max: 30 }),
      fc.boolean(),
      fc.boolean(),
      // empty string not supported when trying to recognize stuff geenrated by this precise arbitrary
      fc.array(fc.fullUnicodeString({ minLength: 1 }), { minLength: 1 }),
    )
    .map(([min, gap, withMin, withMax, patterns]) => ({
      minLength: withMin ? min : undefined,
      maxLength: withMax ? min + gap : undefined,
      patterns,
    }));

  const stringOfBuilder = (extra: Extra) => stringOf(new PatternsArbitrary(extra.patterns), extra);

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(stringOfBuilder, { extraParameters });
  });

  it('should produce values seen as shrinkable without any context', () => {
    assertProduceValuesShrinkableWithoutContext(stringOfBuilder, { extraParameters });
  });

  // assertShrinkProducesSameValueWithoutInitialContext is not fully supported by stringOf
  // Indeed depending how much the source strings overlaps there are possibly many possible ways to build a given string
  // so also many possible ways to shrink it.

  it.each`
    source                             | patterns        | constraints
    ${'.__..' /* not from patterns */} | ${['..', '__']} | ${{}}
    ${'__..' /* not large enough */}   | ${['..', '__']} | ${{ minLength: 3 }}
    ${'__..__..' /* too large */}      | ${['..', '__']} | ${{ maxLength: 3 }}
  `(
    'should not be able to generate $source with fc.stringOf(arb($patterns), $constraints)',
    ({ source, patterns, constraints }) => {
      // Arrange / Act
      const arb = stringOf(new PatternsArbitrary(patterns), constraints);
      const out = arb.canShrinkWithoutContext(source);

      // Assert
      expect(out).toBe(false);
    },
  );

  it.each`
    rawValue                                                                          | patterns
    ${'CccABbBbCcc'}                                                                  | ${['A', 'Bb', 'Ccc']}
    ${'_._.____...'}                                                                  | ${['._', '_...', '_._.', '_..', '.', '.._.', '__.', '....', '..', '.___', '._..', '__', '_.', '___', '.__.', '__._', '._.', '...', '_', '.._', '..._', '.__', '_.._', '_.__', '__..']}
    ${'..'.repeat(50) /* longer than default maxGeneratedLength but ok for shrink */} | ${['..', '__']}
  `('should be able to shrink $rawValue', ({ rawValue, patterns }) => {
    // Arrange
    const arb = stringOf(new PatternsArbitrary(patterns));
    const value = new Value(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});

// Helpers

class PatternsArbitrary extends Arbitrary<string> {
  constructor(readonly patterns: string[]) {
    super();
  }
  generate(mrng: Random): Value<string> {
    return new Value(this.patterns[mrng.nextInt(0, this.patterns.length - 1)], undefined);
  }
  canShrinkWithoutContext(value: unknown): value is string {
    if (typeof value !== 'string') return false;
    return this.patterns.includes(value);
  }
  shrink(value: string): Stream<Value<string>> {
    const patternIndex = this.patterns.indexOf(value);
    if (patternIndex <= 0) {
      return Stream.nil();
    }
    return Stream.of(new Value(this.patterns[0], undefined));
  }
}
