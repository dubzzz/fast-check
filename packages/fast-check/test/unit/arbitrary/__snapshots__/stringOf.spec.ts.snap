// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringOf (integration) should be able to shrink .................................................................................................... 1`] = `
""...................................................................................................."
├> ""
├> ".................................................."
|  ├> ".........................."
|  |  ├> ".............."
|  |  |  ├> "........"
|  |  |  |  ├> "...."
|  |  |  |  |  ├> ".."
|  |  |  |  |  |  └> ""
|  |  |  |  |  └> ".."
|  |  |  |  |     └> ""
|  |  |  |  ├> "......"
|  |  |  |  |  ├> "...."
|  |  |  |  |  |  ├> ".."
|  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  └> ".."
|  |  |  |  |  |     └> ""
|  |  |  |  |  ├> ".."
|  |  |  |  |  |  └> ""
|  |  |  |  |  ├> "...."
|  |  |  |  |  |  ├> ".."
|  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  └> ".."
|  |  |  |  |  |     └> ""
|  |  |  |  |  └> "...."
|  |  |  |  |     ├> ".."
|  |  |  |  |     |  └> ""
|  |  |  |  |     └> ".."
|  |  |  |  |        └> ""
|  |  |  |  ├> ".."
|  |  |  |  |  └> ""
|  |  |  |  ├> "......"
|  |  |  |  |  ├> "...."
|  |  |  |  |  |  ├> ".."
|  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  └> ".."
|  |  |  |  |  |     └> ""
|  |  |  |  |  ├> ".."
|  |  |  |  |  |  └> ""
|  |  |  |  |  ├> "...."
|  |  |  |  |  |  ├> ".."
|  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  └> ".."
|  |  |  |  |  |     └> ""
|  |  |  |  |  └> "...."
|  |  |  |  |     ├> ".."
|  |  |  |  |     |  └> ""
|  |  |  |  |     └> ".."
|  |  |  |  |        └> ""
|  |  |  |  ├> "...."
|  |  |  |  |  ├> ".."
|  |  |  |  |  |  └> ""
|  |  |  |  |  └> ".."
|  |  |  |  |     └> ""
|  |  |  |  ├> "......"
|  |  |  |  |  ├> "...."
|  |  |  |  |  |  ├> ".."
|  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  └> ".."
|  |  |  |  |  |     └> ""
|  |  |  |  |  ├> ".."
|  |  |  |  |  |  └> ""
|  |  |  |  |  ├> "...."
|  |  |  |  |  |  ├> ".."
|  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  └> ".."
|  |  |  |  |  |     └> ""
|  |  |  |  |  └> "...."
|  |  |  |  |     ├> ".."
|  |  |  |  |     |  └> ""
|  |  |  |  |     └> ".."
|  |  |  |  |        └> ""
|  |  |  |  └> "......"
|  |  |  |     ├> "...."
|  |  |  |     |  ├> ".."
|  |  |  |     |  |  └> ""
|  |  |  |     |  └> ".."
|  |  |  |     |     └> ""
|  |  |  |     ├> ".."
|  |  |  |     |  └> ""
|  |  |  |     ├> "...."
|  |  |  |     |  ├> ".."
|  |  |  |     |  |  └> ""
|  |  |  |     |  └> ".."
|  |  |  |     |     └> ""
|  |  |  |     └> "...."
|  |  |  |        ├> ".."
|  |  |  |        |  └> ""
|  |  |  |        └> ".."
|  |  |  |           └> ""
|  |  |  ├> "............"
|  |  |  |  ├> ".........."
|  |  |  |  |  ├> "........"
|  |  |  |  |  |  ├> "...."
|  |  |  |  |  |  |  ├> ".."
|  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  └> ".."
|  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  ├> "......"
|  |  |  |  |  |  |  ├> "...."
|  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`stringOf (integration) should be able to shrink _._.____... 1`] = `
""_._.____..."
├> ""
├> "___..."
|  ├> "..."
|  |  ├> ".."
|  |  |  ├> "."
|  |  |  |  ├> ""
|  |  |  |  └> "._"
|  |  |  |     └> ""
|  |  |  ├> "._."
|  |  |  |  ├> "."
|  |  |  |  |  ├> ""
|  |  |  |  |  └> "._"
|  |  |  |  |     └> ""
|  |  |  |  ├> "._"
|  |  |  |  |  └> ""
|  |  |  |  └> "._._"
|  |  |  |     ├> "._"
|  |  |  |     |  └> ""
|  |  |  |     └> "._"
|  |  |  |        └> ""
|  |  |  ├> "."
|  |  |  |  ├> ""
|  |  |  |  └> "._"
|  |  |  |     └> ""
|  |  |  └> ".._"
|  |  |     ├> "._"
|  |  |     |  └> ""
|  |  |     ├> "._._"
|  |  |     |  ├> "._"
|  |  |     |  |  └> ""
|  |  |     |  └> "._"
|  |  |     |     └> ""
|  |  |     └> "."
|  |  |        ├> ""
|  |  |        └> "._"
|  |  |           └> ""
|  |  ├> "._.."
|  |  |  ├> ".."
|  |  |  |  ├> "."
|  |  |  |  |  ├> ""
|  |  |  |  |  └> "._"
|  |  |  |  |     └> ""
|  |  |  |  ├> "._."
|  |  |  |  |  ├> "."
|  |  |  |  |  |  ├> ""
|  |  |  |  |  |  └> "._"
|  |  |  |  |  |     └> ""
|  |  |  |  |  ├> "._"
|  |  |  |  |  |  └> ""
|  |  |  |  |  └> "._._"
|  |  |  |  |     ├> "._"
|  |  |  |  |     |  └> ""
|  |  |  |  |     └> "._"
|  |  |  |  |        └> ""
|  |  |  |  ├> "."
|  |  |  |  |  ├> ""
|  |  |  |  |  └> "._"
|  |  |  |  |     └> ""
|  |  |  |  └> ".._"
|  |  |  |     ├> "._"
|  |  |  |     |  └> ""
|  |  |  |     ├> "._._"
|  |  |  |     |  ├> "._"
|  |  |  |     |  |  └> ""
|  |  |  |     |  └> "._"
|  |  |  |     |     └> ""
|  |  |  |     └> "."
|  |  |  |        ├> ""
|  |  |  |        └> "._"
|  |  |  |           └> ""
|  |  |  ├> "._"
|  |  |  |  └> ""
|  |  |  ├> "._."
|  |  |  |  ├> "."
|  |  |  |  |  ├> ""
|  |  |  |  |  └> "._"
|  |  |  |  |     └> ""
|  |  |  |  ├> "._"
|  |  |  |  |  └> ""
|  |  |  |  └> "._._"
|  |  |  |     ├> "._"
|  |  |  |     |  └> ""
|  |  |  |     └> "._"
|  |  |  |        └> ""
|  |  |  ├> "._._."
|  |  |  |  ├> "._."
|  |  |  |  |  ├> "."
|  |  |  |  |  |  ├> ""
|  |  |  |  |  |  └> "._"
|  |  |  |  |  |     └> ""
|  |  |  |  |  ├> "._"
|  |  |  |  |  |  └> ""
|  |  |  |  |  └> "._._"
|  |  |  |  |     ├> "._"
|  |  |  |  |     |  └> ""
|  |  |  |  |     └> "._"
|  |  |  |  |        └> ""
|  |  |  |  ├> "._"
|  |  |  |  |  └> ""
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`stringOf (integration) should be able to shrink CccABbBbCcc 1`] = `
""CccABbBbCcc"
├> ""
├> "BbBbCcc"
|  ├> "BbCcc"
|  |  ├> "Ccc"
|  |  |  ├> ""
|  |  |  └> "A"
|  |  |     └> ""
|  |  ├> "ACcc"
|  |  |  ├> "Ccc"
|  |  |  |  ├> ""
|  |  |  |  └> "A"
|  |  |  |     └> ""
|  |  |  ├> "A"
|  |  |  |  └> ""
|  |  |  └> "AA"
|  |  |     ├> "A"
|  |  |     |  └> ""
|  |  |     └> "A"
|  |  |        └> ""
|  |  ├> "Bb"
|  |  |  ├> ""
|  |  |  └> "A"
|  |  |     └> ""
|  |  └> "BbA"
|  |     ├> "A"
|  |     |  └> ""
|  |     ├> "AA"
|  |     |  ├> "A"
|  |     |  |  └> ""
|  |     |  └> "A"
|  |     |     └> ""
|  |     └> "Bb"
|  |        ├> ""
|  |        └> "A"
|  |           └> ""
|  ├> "ABbCcc"
|  |  ├> "BbCcc"
|  |  |  ├> "Ccc"
|  |  |  |  ├> ""
|  |  |  |  └> "A"
|  |  |  |     └> ""
|  |  |  ├> "ACcc"
|  |  |  |  ├> "Ccc"
|  |  |  |  |  ├> ""
|  |  |  |  |  └> "A"
|  |  |  |  |     └> ""
|  |  |  |  ├> "A"
|  |  |  |  |  └> ""
|  |  |  |  └> "AA"
|  |  |  |     ├> "A"
|  |  |  |     |  └> ""
|  |  |  |     └> "A"
|  |  |  |        └> ""
|  |  |  ├> "Bb"
|  |  |  |  ├> ""
|  |  |  |  └> "A"
|  |  |  |     └> ""
|  |  |  └> "BbA"
|  |  |     ├> "A"
|  |  |     |  └> ""
|  |  |     ├> "AA"
|  |  |     |  ├> "A"
|  |  |     |  |  └> ""
|  |  |     |  └> "A"
|  |  |     |     └> ""
|  |  |     └> "Bb"
|  |  |        ├> ""
|  |  |        └> "A"
|  |  |           └> ""
|  |  ├> "A"
|  |  |  └> ""
|  |  ├> "ACcc"
|  |  |  ├> "Ccc"
|  |  |  |  ├> ""
|  |  |  |  └> "A"
|  |  |  |     └> ""
|  |  |  ├> "A"
|  |  |  |  └> ""
|  |  |  └> "AA"
|  |  |     ├> "A"
|  |  |     |  └> ""
|  |  |     └> "A"
|  |  |        └> ""
|  |  ├> "AACcc"
|  |  |  ├> "ACcc"
|  |  |  |  ├> "Ccc"
|  |  |  |  |  ├> ""
|  |  |  |  |  └> "A"
|  |  |  |  |     └> ""
|  |  |  |  ├> "A"
|  |  |  |  |  └> ""
|  |  |  |  └> "AA"
|  |  |  |     ├> "A"
|  |  |  |     |  └> ""
|  |  |  |     └> "A"
|  |  |  |        └> ""
|  |  |  ├> "A"
|  |  |  |  └> ""
|  |  |  ├> "ACcc"
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
